spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/DlqExpiryTests.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.amqp.rabbit.listener;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.jupiter.api.Test;
+
+import org.springframework.amqp.core.Message;
+import org.springframework.amqp.core.Queue;
+import org.springframework.amqp.core.QueueBuilder;
+import org.springframework.amqp.rabbit.annotation.EnableRabbit;
+import org.springframework.amqp.rabbit.annotation.RabbitListener;
+import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory;
+import org.springframework.amqp.rabbit.connection.CachingConnectionFactory;
+import org.springframework.amqp.rabbit.core.RabbitAdmin;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.amqp.rabbit.junit.RabbitAvailable;
+import org.springframework.amqp.rabbit.junit.RabbitAvailableCondition;
+import org.springframework.amqp.support.converter.MessageConversionException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
+
+/**
+ * @author Gary Russell
+ * @since 2.1
+ *
+ */
+@RabbitAvailable
+@SpringJUnitConfig
+public class DlqExpiryTests {
+
+	@Autowired
+	private Config config;
+
+	@Test
+	public void testExpiredDies() throws Exception {
+		this.config.template().convertAndSend("test.expiry.main", "foo");
+		assertThat(this.config.latch.await(10, TimeUnit.SECONDS)).isTrue();
+		Thread.sleep(300);
```

Comment:
This looks suspicious.
As experience shows we are going to have a race condition sporadically.

Why a single `latch` not enough to verify this test-case?

Shouldn't we check instead that we get an `ImmediateAcknowledgeAmqpException` and there is no more messages in the `test.expiry.main` and stop container, of course before verification. 

Well, if you can convince me that this is OK what you have done so far, we can proceed with merging.

Thanks

Reply:
Because it's specifically testing that the count doesn't increase after the latch has counted down.

Running the test with the old code fails.

While it is true that we could get a "false" successful test because of a busy CI server, it would be a false positive, not a false negative (which is what we need to avoid).